open-banking-uk/jws-header-processing/jws-response-header-generation-handler/src/main/java/org/wso2/financial/services/apim/mediation/policies/jws/header/processing/handler/JwsResponseHeaderHandler.java

```diff
@@ -0,0 +1,421 @@
+
+        return true;
+
+    }
+
+    /**
+     * Handle request message going out from the engine.
+     *
+     * @param messageContext outgoing request message context
+     * @return whether mediation flow should continue
+     */
+    @Override
+    public boolean handleRequestOutFlow(MessageContext messageContext) {
+
+        return true;
+    }
+
+    /**
+     * Handle response message coming into the engine.
+     *
+     * @param messageContext incoming response message context
+     * @return whether mediation flow should continue
+     */
+    @Override
+    public boolean handleResponseInFlow(MessageContext messageContext) {
+
+        setProperties(messageContext);
+        return appendJwsSignatureToResponse(messageContext);
+    }
+
+    /**
+     * Handle response message going out from the engine.
+     *
+     * @param messageContext outgoing response message context
+     * @return whether mediation flow should continue
+     */
+    @Override
+    public boolean handleResponseOutFlow(MessageContext messageContext) {
+
+        setProperties(messageContext);
+        org.apache.axis2.context.MessageContext axis2MC =
+                ((Axis2MessageContext) messageContext).getAxis2MessageContext();
+        Map<String, String> headers = (Map<String, String>)
+                axis2MC.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+        if (messageContext.getEnvelope() != null && messageContext.getEnvelope().getBody() != null &&
+                StringUtils.contains(messageContext.getEnvelope().getBody().toString(),
+                        "Schema validation failed")) {
+            // Add jws header for schema errors, This is due to schema validation happens after responseInFlow.
+            // So we need to regenerate the jws for schema validation error responses.
+            return appendJwsSignatureToResponse(messageContext);
+        } else if (headers.containsKey(getJwSignatureHeaderName()) && headers.get(getJwSignatureHeaderName()) != null) {
+            return true;
+        } else {
+            // Add jws header, if it's not added yet.
+            return appendJwsSignatureToResponse(messageContext);
+        }
+    }
+
+    /**
+     * Method to append Jws Signature to the response.
+     *
+     * @param messageContext response/request message context.
+     * @return jws signature response is successfully appended.
+     */
+    private boolean appendJwsSignatureToResponse(MessageContext messageContext) {
+
+        try {
+            boolean applicable = isApplicable(messageContext);
+            if (!applicable) {
+                log.debug("Signature generation is not applicable for this response");
+                return true;
+            } else {
+                log.debug("Generating signature for the response");
+            }
+        } catch (RuntimeException e) {
+            log.debug("Internal Server Error, Unable to append jws signature", e);
```

Comment:
isn't this log error

Reply:
Fixed with eef7dc63510ee9132999de74b167cf8d3c7c1e8a